std/arith.asm

```diff
@@ -161,18 +188,39 @@ machine Arith(CLK32_31, operation_id){
 		- shift_right(y2f, 16)(nr)
 		- y3f(nr)
 	)();
+
+    // Binary selectors for the equations that are activated. Determined from the operation ID via bit-decomposition.
+    // Note that there are only 4 selectors because equation 4 is activated iff. equation 3 is activated, so we can
+    // re-use the same selector.
+    pol commit selEq[4];
+    // Note that this is not necessary, because the operation ID is already constant within the block
+    // array::map(selEq, |e| unchanged_until(e, CLK32[31]));
+    array::map(selEq, |c| force_bool(c));
+    sum(4, |i| 2 ** i * selEq[i]) = operation_id;
     
     // Note that Polygon uses a single 22-Bit column. However, this approach allows for a lower degree (2**16)
     // while still preventing overflows: The 32-bit carry gets added to 32 16-Bit values, which can't overflow
     // the Goldilocks field.
-    pol witness carry_low, carry_high;
-    { carry_high } in { BYTE2 };
-    { carry_low } in { BYTE2 };
-
-    pol carry = carry_high * 2**16 + carry_low;
+    pol witness carry_low[3], carry_high[3];
+    { carry_low[0] } in { BYTE2 };
+    { carry_low[1] } in { BYTE2 };
+    { carry_low[2] } in { BYTE2 };
+    { carry_high[0] } in { BYTE2 };
+    { carry_high[1] } in { BYTE2 };
+    { carry_high[2] } in { BYTE2 };
+
+    // Carries can be any integer in the range [-2**31, 2**31 - 1)
+    pol carry0 = carry_high[0] * 2**16 + carry_low[0] - 2 ** 31;
+    pol carry1 = carry_high[1] * 2**16 + carry_low[1] - 2 ** 31;
+    pol carry2 = carry_high[2] * 2**16 + carry_low[2] - 2 ** 31;
+    let carry = [carry0, carry1, carry2];
     
-    carry * CLK32[0] = 0;
+    array::map(carry, |c| c * CLK32[0] == 0);
+
+    // TODO: Somehow witgen fails if a carry is unconstrained
+    selEq[0] * carry[1] = 0;
+    selEq[0] * carry[2] = 0;
 
-    let eq0_sum = sum(32, |i| eq0(i) * CLK32[i]);
-    carry + eq0_sum = carry' * 2**16;
+    col eq0_sum = sum(32, |i| eq0(i) * CLK32[i]);
```

Comment:
Can't be inlined, because otherwise the degree of the expression is too large and eStark fails to generate a verifying proof.

Reply:
We should fix that.

Reply:
Is it a proper intermediate column? Might be fixed by one of my PRs.

Reply:
I'm not sure if we're talking about the same thing. I'm talking about getting rid of this line and writing the next one as:
`selEq[0] * (sum(32, |i| eq0(i) * CLK32[i]) + carry[0]) = selEq[0] * carry[0]' * 2**16;`

To fix that, we'd have to automatically turn high-degree expressions into lower degree expressions by introducing intermediate polynomials, which is what I'm doing manually here. Arguably it's also more readable this way. I just put the comment because I originally inlined this and as a result the eStark proof didn't verify...

Reply:
Oh interesting. I thought we already do that.